File: backend/app/services/export_service.py
import json
from pathlib import Path
from datetime import datetime, timezone
from app.models.meeting import Meeting
from app.pipeline.summarizer import SummaryResult
from app.pipeline.transcriber import TranscriptionResult


OUTPUT_MD_TEMPLATE = """# {title}

## 基本信息
- 日期：{date}
- 时长：{duration}
- 原始文件：{filename}

## 会议摘要
{summary}

## 关键讨论点
{key_points}

## 决策记录
{decisions}

## 行动项
{action_items}

## 标签
{tags}

---

> 由 MeetingSum 自动生成
"""


def _fmt_duration(seconds: int | None) -> str:
    if seconds is None:
        return "--:--:--"
    h = seconds // 3600
    m = (seconds % 3600) // 60
    s = seconds % 60
    return f"{h:02d}:{m:02d}:{s:02d}"


def _format_key_points(key_points: list[dict]) -> str:
    if not key_points:
        return "_无_"
    lines = []
    for i, kp in enumerate(key_points, 1):
        importance = kp.get("importance", "")
        badge = {"high": "【重要】", "medium": "", "low": ""}.get(importance, "")
        lines.append(f"{i}. **{kp.get('topic', '议题')}** {badge}")
        lines.append(f"   {kp.get('content', '')}")
    return "\n".join(lines)


def _format_decisions(decisions: list[dict]) -> str:
    if not decisions:
        return "_无_"
    lines = []
    for d in decisions:
        proposer = d.get("proposer") or "未指定"
        lines.append(f"- [ ] {d.get('content', '')} — 提出人：{proposer}")
    return "\n".join(lines)


def _format_action_items(items: list[dict]) -> str:
    if not items:
        return "_无_"
    lines = []
    for item in items:
        assignee = item.get("assignee") or "待指定"
        deadline = item.get("deadline") or "待定"
        lines.append(f"- [ ] {item.get('content', '')} — 负责人：{assignee} — 截止日期：{deadline}")
    return "\n".join(lines)


def generate_markdown(
    meeting: Meeting,
    summary: SummaryResult,
    output_dir: str,
) -> str:
    """Generate a Markdown summary document and return the file path."""
    output_path = Path(output_dir) / f"{meeting.id}_summary.md"

    date_str = meeting.created_at.strftime("%Y-%m-%d") if meeting.created_at else datetime.now(timezone.utc).strftime("%Y-%m-%d")

    content = OUTPUT_MD_TEMPLATE.format(
        title=summary.title or meeting.title or "未命名会议",
        date=date_str,
        duration=_fmt_duration(meeting.duration_seconds),
        filename=Path(meeting.original_file or "").name or "未知",
        summary=summary.summary,
        key_points=_format_key_points(summary.key_points),
        decisions=_format_decisions(summary.decisions),
        action_items=_format_action_items(summary.action_items),
        tags=", ".join(summary.tags) if summary.tags else "_无_",
    )

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(content, encoding="utf-8")
    return str(output_path)


def generate_transcript_markdown(
    meeting: Meeting,
    transcription: TranscriptionResult,
    output_dir: str,
) -> str:
    """Generate a Markdown file of the full transcript with timestamps."""
    output_path = Path(output_dir) / f"{meeting.id}_transcript.md"

    lines = [
        f"# 完整转录文本 — {meeting.title or '未命名会议'}",
        "",
        f"日期：{meeting.created_at.strftime('%Y-%m-%d') if meeting.created_at else ''}",
        f"时长：{_fmt_duration(meeting.duration_seconds)}",
        "",
        "---",
        "",
        transcription.text_with_timestamps,
        "",
        "> 由 MeetingSum 自动生成",
    ]

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text("\n".join(lines), encoding="utf-8")
    return str(output_path)


def generate_docx(
    meeting: Meeting,
    summary: SummaryResult,
    output_dir: str,
) -> str:
    """Generate a Word (.docx) document and return the file path."""
    from docx import Document
    from docx.shared import Pt, Inches

    output_path = Path(output_dir) / f"{meeting.id}_summary.docx"
    output_path.parent.mkdir(parents=True, exist_ok=True)

    doc = Document()

    style = doc.styles["Normal"]
    font = style.font
    font.name = "Arial"
    font.size = Pt(11)

    doc.add_heading(summary.title or meeting.title or "未命名会议", level=1)

    doc.add_heading("基本信息", level=2)
    date_str = meeting.created_at.strftime("%Y-%m-%d") if meeting.created_at else ""
    doc.add_paragraph(f"日期：{date_str}")
    doc.add_paragraph(f"时长：{_fmt_duration(meeting.duration_seconds)}")
    doc.add_paragraph(f"原始文件：{Path(meeting.original_file or '').name or '未知'}")

    doc.add_heading("会议摘要", level=2)
    doc.add_paragraph(summary.summary)

    doc.add_heading("关键讨论点", level=2)
    for i, kp in enumerate(summary.key_points, 1):
        importance = kp.get("importance", "")
        prefix = "【重要】" if importance == "high" else ""
        doc.add_paragraph(f"{i}. {prefix}{kp.get('topic', '议题')}", style="List Number")
        doc.add_paragraph(kp.get("content", ""))

    doc.add_heading("决策记录", level=2)
    for d in summary.decisions:
        proposer = d.get("proposer") or "未指定"
        doc.add_paragraph(f"{d.get('content', '')} — 提出人：{proposer}", style="List Bullet")

    doc.add_heading("行动项", level=2)
    for item in summary.action_items:
        assignee = item.get("assignee") or "待指定"
        deadline = item.get("deadline") or "待定"
        doc.add_paragraph(
            f"{item.get('content', '')} — 负责人：{assignee} — 截止日期：{deadline}",
            style="List Bullet",
        )

    if summary.tags:
        doc.add_heading("标签", level=2)
        doc.add_paragraph(", ".join(summary.tags))

    doc.add_paragraph("")
    doc.add_paragraph("由 MeetingSum 自动生成")

    doc.save(str(output_path))
    return str(output_path)


def export_all(
    meeting: Meeting,
    summary: SummaryResult,
    transcription: TranscriptionResult,
    output_dir: str,
) -> dict[str, str]:
    """Generate all output documents. Returns a dict of format -> path."""
    results = {}

    md_path = generate_markdown(meeting, summary, output_dir)
    results["md"] = md_path

    transcript_path = generate_transcript_markdown(meeting, transcription, output_dir)
    results["transcript"] = transcript_path

    try:
        docx_path = generate_docx(meeting, summary, output_dir)
        results["docx"] = docx_path
    except Exception:
        pass

    return results

File: Java_backend/src/main/java/com/meetingsum/pipeline/ExportService.java
package com.meetingsum.pipeline;

import com.meetingsum.model.dto.SummaryData;
import com.meetingsum.model.entity.Meeting;
import com.meetingsum.pipeline.TranscriberService.WhisperResult;
import com.meetingsum.pipeline.TranscriberService.WhisperResult.Segment;
import org.apache.poi.xwpf.usermodel.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDateTime;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@Component
public class ExportService {

    private static final Logger log = LoggerFactory.getLogger(ExportService.class);

    private static final String MD_TEMPLATE = """
            # %s

            ## 基本信息
            - 日期：%s
            - 时长：%s
            - 原始文件：%s

            ## 会议摘要
            %s

            ## 关键讨论点
            %s

            ## 决策记录
            %s

            ## 行动项
            %s

            ## 标签
            %s

            ---

            > 由 MeetingSum 自动生成
            """;

    public String generateMarkdown(Meeting meeting, SummaryData summary, String outputDir) throws IOException {
        Path outputPath = Path.of(outputDir).resolve(meeting.getId() + "_summary.md");
        outputPath.getParent().toFile().mkdirs();

        String dateStr = meeting.getCreatedAt() != null
                ? meeting.getCreatedAt().toLocalDate().toString()
                : LocalDateTime.now().toLocalDate().toString();

        String filename = meeting.getOriginalFile() != null
                ? Path.of(meeting.getOriginalFile()).getFileName().toString()
                : "unknown";

        String content = String.format(MD_TEMPLATE,
                summary.title() != null && !summary.title().isBlank() ? summary.title() : (meeting.getTitle() != null ? meeting.getTitle() : "未命名会议"),
                dateStr,
                fmtDuration(meeting.getDurationSeconds()),
                filename,
                summary.summary(),
                formatKeyPoints(summary.keyPoints()),
                formatDecisions(summary.decisions()),
                formatActionItems(summary.actionItems()),
                summary.tags() != null && !summary.tags().isEmpty() ? String.join(", ", summary.tags()) : "_无_"
        );

        Files.writeString(outputPath, content, StandardCharsets.UTF_8);
        return outputPath.toString();
    }

    public String generateTranscriptMarkdown(Meeting meeting, WhisperResult transcription, String outputDir) throws IOException {
        Path outputPath = Path.of(outputDir).resolve(meeting.getId() + "_transcript.md");
        outputPath.getParent().toFile().mkdirs();

        StringBuilder sb = new StringBuilder();
        sb.append("# 完整转录文本 — ").append(meeting.getTitle() != null ? meeting.getTitle() : "未命名会议").append("\n\n");
        sb.append("日期：").append(meeting.getCreatedAt() != null ? meeting.getCreatedAt().toLocalDate().toString() : "").append("\n");
        sb.append("时长：").append(fmtDuration(meeting.getDurationSeconds())).append("\n\n");
        sb.append("---\n\n");

        for (Segment seg : transcription.segments()) {
            String ts = String.format("[%s -> %s]", fmtTime(seg.start()), fmtTime(seg.end()));
            String prefix = seg.speaker() != null ? "[" + seg.speaker() + "] " : "";
            sb.append(ts).append(" ").append(prefix).append(seg.text()).append("\n");
        }

        sb.append("\n> 由 MeetingSum 自动生成\n");
        Files.writeString(outputPath, sb.toString(), StandardCharsets.UTF_8);
        return outputPath.toString();
    }

    public String generateDocx(Meeting meeting, SummaryData summary, String outputDir) throws IOException {
        Path outputPath = Path.of(outputDir).resolve(meeting.getId() + "_summary.docx");
        outputPath.getParent().toFile().mkdirs();

        try (XWPFDocument doc = new XWPFDocument()) {
            String title = summary.title() != null && !summary.title().isBlank()
                    ? summary.title()
                    : (meeting.getTitle() != null ? meeting.getTitle() : "未命名会议");

            doc.createParagraph().createRun().setText(title);
            doc.createParagraph().createRun().setBold(true);

            addHeading(doc, "基本信息", 2);
            String dateStr = meeting.getCreatedAt() != null
                    ? meeting.getCreatedAt().toLocalDate().toString()
                    : "";
            doc.createParagraph().createRun().setText("日期：" + dateStr);
            doc.createParagraph().createRun().setText("时长：" + fmtDuration(meeting.getDurationSeconds()));
            String filename = meeting.getOriginalFile() != null
                    ? Path.of(meeting.getOriginalFile()).getFileName().toString()
                    : "未知";
            doc.createParagraph().createRun().setText("原始文件：" + filename);

            addHeading(doc, "会议摘要", 2);
            doc.createParagraph().createRun().setText(summary.summary() != null ? summary.summary() : "");

            addHeading(doc, "关键讨论点", 2);
            if (summary.keyPoints() != null) {
                for (int i = 0; i < summary.keyPoints().size(); i++) {
                    var kp = summary.keyPoints().get(i);
                    String prefix = "high".equals(kp.importance()) ? "【重要】" : "";
                    XWPFParagraph p = doc.createParagraph();
                    p.createRun().setText((i + 1) + ". " + prefix + (kp.topic() != null ? kp.topic() : "议题"));
                    doc.createParagraph().createRun().setText(kp.content() != null ? kp.content() : "");
                }
            }

            addHeading(doc, "决策记录", 2);
            if (summary.decisions() != null) {
                for (var d : summary.decisions()) {
                    String proposer = d.proposer() != null ? d.proposer() : "未指定";
                    doc.createParagraph().createRun().setText(
                            (d.content() != null ? d.content() : "") + " — 提出人：" + proposer
                    );
                }
            }

            addHeading(doc, "行动项", 2);
            if (summary.actionItems() != null) {
                for (var item : summary.actionItems()) {
                    String assignee = item.assignee() != null ? item.assignee() : "待指定";
                    String deadline = item.deadline() != null ? item.deadline() : "待定";
                    doc.createParagraph().createRun().setText(
                            (item.content() != null ? item.content() : "") + " — 负责人：" + assignee + " — 截止日期：" + deadline
                    );
                }
            }

            if (summary.tags() != null && !summary.tags().isEmpty()) {
                addHeading(doc, "标签", 2);
                doc.createParagraph().createRun().setText(String.join(", ", summary.tags()));
            }

            doc.createParagraph();
            doc.createParagraph().createRun().setText("由 MeetingSum 自动生成");

            try (FileOutputStream fos = new FileOutputStream(outputPath.toFile())) {
                doc.write(fos);
            }
        }

        return outputPath.toString();
    }

    public Map<String, String> exportAll(Meeting meeting, SummaryData summary, WhisperResult transcription, String outputDir) throws IOException {
        Map<String, String> results = new LinkedHashMap<>();

        String mdPath = generateMarkdown(meeting, summary, outputDir);
        results.put("md", mdPath);

        String transcriptPath = generateTranscriptMarkdown(meeting, transcription, outputDir);
        results.put("transcript", transcriptPath);

        try {
            String docxPath = generateDocx(meeting, summary, outputDir);
            results.put("docx", docxPath);
        } catch (Exception e) {
            log.warn("DOCX generation failed, continuing without it", e);
        }

        return results;
    }

    private void addHeading(XWPFDocument doc, String text, int level) {
        XWPFParagraph p = doc.createParagraph();
        XWPFRun run = p.createRun();
        run.setText(text);
        run.setBold(true);
        run.setFontSize(level == 1 ? 18 : 14);
    }

    private String fmtDuration(Integer seconds) {
        if (seconds == null) return "--:--:--";
        int h = seconds / 3600;
        int m = (seconds % 3600) / 60;
        int s = seconds % 60;
        return String.format("%02d:%02d:%02d", h, m, s);
    }

    private String fmtTime(double seconds) {
        int h = (int) (seconds / 3600);
        int m = (int) ((seconds % 3600) / 60);
        int s = (int) (seconds % 60);
        return String.format("%02d:%02d:%02d", h, m, s);
    }

    private String formatKeyPoints(List<SummaryData.KeyPoint> keyPoints) {
        if (keyPoints == null || keyPoints.isEmpty()) return "_无_";
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < keyPoints.size(); i++) {
            var kp = keyPoints.get(i);
            String badge = "high".equals(kp.importance()) ? "【重要】" : "";
            sb.append(i + 1).append(". **").append(kp.topic() != null ? kp.topic() : "议题").append("** ").append(badge).append("\n");
            sb.append("   ").append(kp.content() != null ? kp.content() : "").append("\n");
        }
        return sb.toString();
    }

    private String formatDecisions(List<SummaryData.Decision> decisions) {
        if (decisions == null || decisions.isEmpty()) return "_无_";
        StringBuilder sb = new StringBuilder();
        for (var d : decisions) {
            String proposer = d.proposer() != null ? d.proposer() : "未指定";
            sb.append("- [ ] ").append(d.content() != null ? d.content() : "").append(" — 提出人：").append(proposer).append("\n");
        }
        return sb.toString();
    }

    private String formatActionItems(List<SummaryData.ActionItem> items) {
        if (items == null || items.isEmpty()) return "_无_";
        StringBuilder sb = new StringBuilder();
        for (var item : items) {
            String assignee = item.assignee() != null ? item.assignee() : "待指定";
            String deadline = item.deadline() != null ? item.deadline() : "待定";
            sb.append("- [ ] ").append(item.content() != null ? item.content() : "")
                    .append(" — 负责人：").append(assignee)
                    .append(" — 截止日期：").append(deadline).append("\n");
        }
        return sb.toString();
    }
}
